Implement TH2Poly in DQM Services and add GetArray support

ywkao opened the pull request:
#### PR description:

This PR includes both the TH2Poly implementation from https://github.com/cms-sw/cmssw/pull/41932 and the additional GetArray support for TH2Poly histograms.

The DQMStoreStats::calcstats() method handles bin contents of DQM monitor elements by using the GetArray() method, but the newly integrated TH2Poly histogram type lacks this method.

This PR provides a solution by implementing a custom GetTH2PolyArray() function that extracts bin contents from TH2Poly objects and formats them appropriately for the DQM framework.

#### PR validation:

The GetArray() functionality has been checked using independent C++ code. Additional validation within the CMSSW framework is ongoing.

Note: This PR could potentially supersede #41932 as it contains all those changes plus the GetArray support, as discussed in the PR comments of #41932.

ywkao committed:
Implement TH2Poly in DQM Services for HGCal DQM

ywkao committed:
Apply changes from code quality checks

ywkao committed:
Add a patch based on makortel's comments

ywkao committed:
Implement TH2Poly support in DQMServices
- Add TH2Poly support in DQMServices/Core
- Implement TH2Poly in DQMServices/Components
- Add TH2Poly support in MEtoEDMFormat.h

ywkao committed:
Extend DQMServices testing for TH2Poly
- Add TH2Poly support in FwkIO and StreamerIO
- Add TH2Poly in DQMServices/Demo/test

ywkao committed:
Apply code-format patch

ywkao committed:
Fix IndexError in the DQM unit test for TH2Poly

ywkao committed:
Fix TypeError in `DQMServices/Demo/test/dqmiodumpentries.py`. Omit DeprecationWarning of `getattr()`.

| ROOT version | `getattr(f,"x")` | `f["x"]`    |
| ------------ | ---------------- | ----------- |
| v6.30        | works            | Not working |
| v6.35        | works*           | works       |
| v6.32        | works*           | works       |

* This will display deprecation warning message

ywkao committed:
Fix typo in DataFormats/Histograms/src/classes_def.xml

ywkao committed:
Fix TH2Poly handling in DQM Monitor Elements

- Add support for GetNcells() to identify empty TH2Poly objects
- Update Reset() functionality for DQM Monitor Elements
- Fix bin accumulation issue in TH2Poly addition
- Add more support for TH2Poly::AddBin() methods
- Add unit tests with polygonal bins

ywkao committed:
apply code checks and format recommendation

ywkao committed:
Fix MonitorElement::getNcells() dimension handling

- Add proper dimension checks based on histogram type (TH1/TH2/TH3)
- Match dimension handling pattern used in MonitorElement::Reset() method
- Fix FastTimerService fatal error related to incorrect dimension access

ywkao committed:
Remove commented TH2Poly code as requested by reviewers

cmsbuild commented:
cms-bot internal usage<!-- bot cache: {"commits":{"07470c10e048ae5a907d2a3870a90044a2a889b2":{"files":["DQMServices/FwkIO/plugins/DQMRootSource.cc"],"squashed":false,"time":1734590165},"419cb6f7c4bd6b461e3f3984b13ab1ea1e124c9c":{"files":["DQMServices/Core/interface/DQMStore.h"],"squashed":false,"time":1733201609},"4fed83bdbd308ec4dd98827ca51b77f2c289c2a9":{"files":["DQMServices/Components/plugins/DQMStoreStats.cc"],"squashed":false,"time":1740714122},"5284c46a015ef3bd8351b577d069962aa462aee2":{"files":["DQMServices/Core/interface/DQMStore.h","DQMServices/Core/src/MonitorElement.cc","DataFormats/Histograms/interface/MonitorElementCollection.h"],"squashed":false,"time":1733201611},"59617bc0b430fe35b5fd212706a3d37984fc841d":{"files":["DataFormats/Histograms/src/classes_def.xml"],"squashed":false,"time":1733237119},"7755770ac837d971e6f5068ca7124f0d42496f7c":{"files":["DQMServices/Core/src/MonitorElement.cc"],"squashed":false,"time":1738123744},"7d85734d22696b37867982b51ed0ebe522ef337a":{"files":["DQMServices/Demo/test/TestDQMEDAnalyzer.cc","DQMServices/Demo/test/dqmiodumpentries.py","DQMServices/Demo/test/runtests.sh","DQMServices/FwkIO/plugins/DQMRootOutputModule.cc","DQMServices/FwkIO/plugins/DQMRootSource.cc","DQMServices/FwkIO/plugins/format.h","DQMServices/StreamerIO/plugins/DQMProtobufReader.cc","DataFormats/Histograms/src/classes_def.xml"],"squashed":false,"time":1733201878},"8041853ab76acb377742861567b7129ce8e155d4":{"files":["DQMServices/Core/interface/DQMStore.h","DQMServices/Core/interface/MonitorElement.h","DQMServices/Core/src/MonitorElement.cc","DataFormats/Histograms/interface/MEtoEDMFormat.h","DataFormats/Histograms/interface/MonitorElementCollection.h"],"squashed":false,"time":1733201607},"c0af007890f85e74a05905304886c796839f9413":{"files":["DQMServices/Core/interface/MonitorElement.h","DQMServices/Core/src/MonitorElement.cc","DQMServices/Demo/test/TestDQMEDAnalyzer.cc","DQMServices/Demo/test/runtests.sh","DataFormats/Histograms/interface/MEtoEDMFormat.h","DataFormats/Histograms/test/metoedmformat_t.cc"],"squashed":false,"time":1737619656},"cca126b0fe0bc5893dfc5c3977416359da0297da":{"files":["DQMServices/Components/scripts/dqmiolistmes.py","DQMServices/Demo/test/runtests.sh"],"squashed":false,"time":1733201890},"eb416480b2cf26248858413797953d634ea4deda":{"files":["DQMServices/Demo/test/dqmiodumpentries.py"],"squashed":false,"time":1733229404},"f750a3337adf1719ee2dc2f573bf6fffebf7e9d3":{"files":["DQMServices/Core/interface/DQMStore.h","DQMServices/Core/interface/MonitorElement.h","DQMServices/Demo/test/TestDQMEDAnalyzer.cc","DataFormats/Histograms/interface/MEtoEDMFormat.h"],"squashed":false,"time":1733201888},"fb37861aa536c120e63553c38fec73e744bf252f":{"files":["DQMServices/Components/plugins/DQMStoreStats.cc","DQMServices/Components/plugins/EDMtoMEConverter.cc","DQMServices/Components/plugins/MEtoEDMConverter.cc","DQMServices/Components/scripts/dqmiolistmes.py","DQMServices/Core/interface/DQMNet.h","DQMServices/Core/interface/DQMStore.h","DQMServices/Core/src/DQMNet.cc","DQMServices/Core/src/LegacyIOHelper.cc","DataFormats/Histograms/interface/MEtoEDMFormat.h"],"squashed":false,"time":1733201612}},"emoji":{},"last_seen_sha":"4fed83bdbd308ec4dd98827ca51b77f2c289c2a9","signatures":{}} -->

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-47478/43918

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DQMServices/Components/plugins/EDMtoMEConverter.cc  modified in PR(s): #41932
  - File  DQMServices/Components/plugins/MEtoEDMConverter.cc  modified in PR(s): #41932
  - File  DQMServices/Components/scripts/dqmiolistmes.py  modified in PR(s): #41932
  - File  DQMServices/Core/interface/DQMNet.h  modified in PR(s): #41932
  - File  DQMServices/Core/interface/DQMStore.h  modified in PR(s): #41932
  - File  DQMServices/Core/interface/MonitorElement.h  modified in PR(s): #41932
  - File  DQMServices/Core/src/DQMNet.cc  modified in PR(s): #41932, #46247
  - File  DQMServices/Core/src/LegacyIOHelper.cc  modified in PR(s): #41932
  - File  DQMServices/Core/src/MonitorElement.cc  modified in PR(s): #41932
  - File  DQMServices/Demo/test/TestDQMEDAnalyzer.cc  modified in PR(s): #41932
  - File  DQMServices/Demo/test/dqmiodumpentries.py  modified in PR(s): #41932
  - File  DQMServices/Demo/test/runtests.sh  modified in PR(s): #41932
  - File  DQMServices/FwkIO/plugins/DQMRootOutputModule.cc  modified in PR(s): #41932, #44372
  - File  DQMServices/FwkIO/plugins/DQMRootSource.cc  modified in PR(s): #41932, #44372
  - File  DQMServices/FwkIO/plugins/format.h  modified in PR(s): #41932
  - File  DQMServices/StreamerIO/plugins/DQMProtobufReader.cc  modified in PR(s): #41932
  - File  DataFormats/Histograms/interface/MEtoEDMFormat.h  modified in PR(s): #41932
  - File  DataFormats/Histograms/interface/MonitorElementCollection.h  modified in PR(s): #41932
  - File  DataFormats/Histograms/src/classes_def.xml  modified in PR(s): #41932
  - File  DataFormats/Histograms/test/metoedmformat_t.cc  modified in PR(s): #41932

ywkao committed:
Add GetTH2PolyArray implementation for DQMStoreStats

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-47478/43919

ERROR: Build errors found during clang-tidy run.
```
src/DQMServices/Components/plugins/DQMStoreStats.cc:691:7: error: cannot jump from switch statement to this case label [clang-diagnostic-error]
  691 |       case MonitorElement::Kind::TPROFILE2D:
      |       ^
src/DQMServices/Components/plugins/DQMStoreStats.cc:683:13: note: jump bypasses variable initialization
--
src/DQMServices/Components/plugins/DQMStoreStats.cc:701:7: error: cannot jump from switch statement to this case label [clang-diagnostic-error]
  701 |       case MonitorElement::Kind::TH3F:
      |       ^
src/DQMServices/Components/plugins/DQMStoreStats.cc:683:13: note: jump bypasses variable initialization
--
src/DQMServices/Components/plugins/DQMStoreStats.cc:711:7: error: cannot jump from switch statement to this case label [clang-diagnostic-error]
  711 |       default: {
      |       ^
src/DQMServices/Components/plugins/DQMStoreStats.cc:683:13: note: jump bypasses variable initialization
--
gmake: *** [config/SCRAM/GMake/Makefile.coderules:129: code-checks] Error 2
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 2
```

ywkao committed:
Fix switch case braces in DQMStoreStats

Add missing curly braces around TH2Poly case block to properly
scope local variable declarations.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-47478/43920

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DQMServices/Components/plugins/DQMStoreStats.cc  modified in PR(s): #46247
  - File  DQMServices/Components/plugins/EDMtoMEConverter.cc  modified in PR(s): #41932
  - File  DQMServices/Components/plugins/MEtoEDMConverter.cc  modified in PR(s): #41932
  - File  DQMServices/Components/scripts/dqmiolistmes.py  modified in PR(s): #41932
  - File  DQMServices/Core/interface/DQMNet.h  modified in PR(s): #41932
  - File  DQMServices/Core/interface/DQMStore.h  modified in PR(s): #41932
  - File  DQMServices/Core/interface/MonitorElement.h  modified in PR(s): #41932
  - File  DQMServices/Core/src/DQMNet.cc  modified in PR(s): #41932, #46247
  - File  DQMServices/Core/src/LegacyIOHelper.cc  modified in PR(s): #41932
  - File  DQMServices/Core/src/MonitorElement.cc  modified in PR(s): #41932
  - File  DQMServices/Demo/test/TestDQMEDAnalyzer.cc  modified in PR(s): #41932
  - File  DQMServices/Demo/test/dqmiodumpentries.py  modified in PR(s): #41932
  - File  DQMServices/Demo/test/runtests.sh  modified in PR(s): #41932
  - File  DQMServices/FwkIO/plugins/DQMRootOutputModule.cc  modified in PR(s): #41932, #44372
  - File  DQMServices/FwkIO/plugins/DQMRootSource.cc  modified in PR(s): #41932, #44372
  - File  DQMServices/FwkIO/plugins/format.h  modified in PR(s): #41932
  - File  DQMServices/StreamerIO/plugins/DQMProtobufReader.cc  modified in PR(s): #41932
  - File  DataFormats/Histograms/interface/MEtoEDMFormat.h  modified in PR(s): #41932
  - File  DataFormats/Histograms/interface/MonitorElementCollection.h  modified in PR(s): #41932
  - File  DataFormats/Histograms/src/classes_def.xml  modified in PR(s): #41932
  - File  DataFormats/Histograms/test/metoedmformat_t.cc  modified in PR(s): #41932

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-47478/43920/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-47478/43920/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

ywkao committed:
Apply code format check

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-47478/43936

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DQMServices/Components/plugins/DQMStoreStats.cc  modified in PR(s): #46247
  - File  DQMServices/Components/plugins/EDMtoMEConverter.cc  modified in PR(s): #41932
  - File  DQMServices/Components/plugins/MEtoEDMConverter.cc  modified in PR(s): #41932
  - File  DQMServices/Components/scripts/dqmiolistmes.py  modified in PR(s): #41932
  - File  DQMServices/Core/interface/DQMNet.h  modified in PR(s): #41932
  - File  DQMServices/Core/interface/DQMStore.h  modified in PR(s): #41932
  - File  DQMServices/Core/interface/MonitorElement.h  modified in PR(s): #41932
  - File  DQMServices/Core/src/DQMNet.cc  modified in PR(s): #41932, #46247
  - File  DQMServices/Core/src/LegacyIOHelper.cc  modified in PR(s): #41932
  - File  DQMServices/Core/src/MonitorElement.cc  modified in PR(s): #41932
  - File  DQMServices/Demo/test/TestDQMEDAnalyzer.cc  modified in PR(s): #41932
  - File  DQMServices/Demo/test/dqmiodumpentries.py  modified in PR(s): #41932
  - File  DQMServices/Demo/test/runtests.sh  modified in PR(s): #41932
  - File  DQMServices/FwkIO/plugins/DQMRootOutputModule.cc  modified in PR(s): #41932, #44372
  - File  DQMServices/FwkIO/plugins/DQMRootSource.cc  modified in PR(s): #41932, #44372
  - File  DQMServices/FwkIO/plugins/format.h  modified in PR(s): #41932
  - File  DQMServices/StreamerIO/plugins/DQMProtobufReader.cc  modified in PR(s): #41932
  - File  DataFormats/Histograms/interface/MEtoEDMFormat.h  modified in PR(s): #41932
  - File  DataFormats/Histograms/interface/MonitorElementCollection.h  modified in PR(s): #41932
  - File  DataFormats/Histograms/src/classes_def.xml  modified in PR(s): #41932
  - File  DataFormats/Histograms/test/metoedmformat_t.cc  modified in PR(s): #41932

cmsbuild commented:
Pull request #47478 was updated.

mmusich commented:
@cmsbuild, please test

cmsbuild commented:
A new Pull Request was created by  @ywkao for master.

It involves the following packages:

- DQMServices/Components (**dqm**)
- DQMServices/Core (**dqm**)
- DQMServices/Demo (**dqm**)
- DQMServices/FwkIO (**dqm**)
- DQMServices/StreamerIO (**dqm**)
- DataFormats/Histograms (**dqm, core**)


@Dr15Jones, @antoniovagnerini, @makortel, @rseidita, @smuzaffar can you please review it and eventually sign? Thanks.
@barvic, @missirol, @mmusich, @rovere this is something you requested to watch as well.
@antoniovilela, @mandrenguyen, @rappoccio, @sextonkennedy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

antoniovagnerini commented:
+dqm

makortel commented:
+core

Fine with me, even if I personally would have found more clear to me to just include the last 3 commits in https://github.com/cms-sw/cmssw/pull/41932.

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @sextonkennedy, @antoniovilela, @rappoccio, @mandrenguyen (and backports should be raised in the release meeting by the corresponding L2)

makortel commented:
@ywkao Could you change the PR title to include the title of #41932 so that the purpose of this PR will be more clear in the release notes?
